File: 21-code-snippets/connect-and-query.cs
using System;
using Npgsql;

class Program
{
    static void Main(string[] args)
    {
        string connString = args.Length > 0
            ? args[0]
            : "Host=localhost;Database=postgres;Username=postgres";

        try
        {
            // NpgsqlConnection implements IDisposable – using ensures cleanup
            using var conn = new NpgsqlConnection(connString);
            conn.Open();
            Console.WriteLine($"Connected to {conn.Database} on {conn.Host}");

            // Execute a query
            string sql = @"SELECT schemaname, tablename
                           FROM pg_tables
                           WHERE tableowner = current_user
                           LIMIT 10";

            using var cmd = new NpgsqlCommand(sql, conn);
            using var reader = cmd.ExecuteReader();

            // Print column headers
            for (int i = 0; i < reader.FieldCount; i++)
                Console.Write($"{reader.GetName(i),-30}");
            Console.WriteLine();
            Console.WriteLine(new string('-', 60));

            // Iterate rows
            while (reader.Read())
            {
                for (int i = 0; i < reader.FieldCount; i++)
                    Console.Write($"{reader[i],-30}");
                Console.WriteLine();
            }

            Console.WriteLine($"\nRows returned: {reader.RecordsAffected}");
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"Error: {ex.Message}");
            Environment.Exit(1);
        }
    }
}

File: 21-code-snippets/connect-and-query.cpp
#include <iostream>
#include <pqxx/pqxx>

int main(int argc, char* argv[]) {
    try {
        // Build connection string from command line or use default
        std::string conn_str = (argc > 1) ? argv[1] : "dbname=postgres";
        pqxx::connection conn{conn_str};

        // Create a non-transactional query (auto-commit)
        pqxx::nontransaction txn{conn};

        // Execute a query
        pqxx::result res = txn.exec(
            "SELECT schemaname, tablename FROM pg_tables WHERE tableowner = current_user LIMIT 10");

        // Print column headers
        for (int i = 0; i < res.columns(); ++i)
            std::cout << res.column_name(i) << "\t";
        std::cout << "\n" << std::string(40, '-') << "\n";

        // Iterate rows
        for (auto const& row : res) {
            for (auto const& field : row)
                std::cout << field.c_str() << "\t";
            std::cout << "\n";
        }

        // txn commits automatically on destruction (RAII)
    } catch (std::exception const& e) {
        std::cerr << "ERROR: " << e.what() << "\n";
        return 1;
    }
    return 0;
}